torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -138,7 +138,108 @@ def get_tp_parallel_strategy(
     return RowwiseParallel, ColwiseParallel
 
 
-def apply_pipeline_parallelism(model, world_mesh, parallel_dims, job_config: JobConfig):
+class DummyTransformerLayer(torch.nn.Module):
+    def forward(self, input, freqs_cis):
+        return input
+
+
+class TransformerChunk(torch.nn.Module):
```

Comment:
wondering if we can put `DummyTransformerLayer` and `TransformerChunk` to a separate file in this folder (i.e. pp_llama_utils.py) and import them here?

Reply:
I want to kill DummyTransformerLayer actually.

I thought about introducing some kind of `PPModuleList` that can wrap an original ModuleList and delete some layers but preserve fqn's of the non-deleted layers.  Then the forward code can still iterate over its remaining layers, but we dont run into stupid issues like stage1's TP tries to apply PrepareInput hook onto layers.0 which is a "DummyModule" but actually that hook will run. (This is a real issue that i hit, but its fixed now since your change to remove PrepareInput hook from layers.0).

Reply:
The same issue has been discussed when doing EP. `ModuleDict` was one potential solution at the time. But that also means that the model needs to use `ModuleDict`.

Reply:
forgot about this. it solves my problem. thank you!